pkg/autoupdate/tuf/library_manager.go

```diff
@@ -320,53 +246,246 @@ func (ulm *updateLibraryManager) tidyUpdateLibrary(binary autoupdatableBinary, c
-
 	// Loop through, looking at the most recent versions first, and remove all once we hit nonCurrentlyRunningVersionsKept valid executables
 	nonCurrentlyRunningVersionsKept := 0
 	for i := len(versionsInLibrary) - 1; i >= 0; i -= 1 {
 		// Always keep the current running executable
-		if versionsInLibrary[i].Original() == currentRunningVersion {
+		if versionsInLibrary[i] == currentRunningVersion {
 			continue
 		}
 
 		// If we've already hit the number of versions to keep, then start to remove the older ones.
 		// We want to keep numberOfVersionsToKeep total, saving a spot for the currently running version.
 		if nonCurrentlyRunningVersionsKept >= numberOfVersionsToKeep-1 {
-			ulm.removeUpdate(binary, versionsInLibrary[i].Original())
+			ulm.removeUpdate(binary, versionsInLibrary[i])
+			continue
+		}
+
+		nonCurrentlyRunningVersionsKept += 1
+	}
+}
+
+// sortedVersionsInLibrary looks through the update library for the given binary to validate and sort all
+// available versions. It returns a sorted list of the valid versions, a list of invalid versions, and
+// an error only when unable to glob for versions.
+func (ulm *updateLibraryManager) sortedVersionsInLibrary(binary autoupdatableBinary) ([]string, []string, error) {
+	rawVersionsInLibrary, err := filepath.Glob(filepath.Join(ulm.updatesDirectory(binary), "*"))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not glob for updates in library: %w", err)
+	}
+
+	versionsInLibrary := make([]*semver.Version, 0)
+	invalidVersions := make([]string, 0)
+	for _, rawVersionWithPath := range rawVersionsInLibrary {
+		rawVersion := filepath.Base(rawVersionWithPath)
+		v, err := semver.NewVersion(rawVersion)
+		if err != nil {
+			invalidVersions = append(invalidVersions, rawVersion)
 			continue
 		}
 
-		// Only keep good executables
-		versionDir := filepath.Join(ulm.updatesDirectory(binary), versionsInLibrary[i].Original())
+		versionDir := filepath.Join(ulm.updatesDirectory(binary), rawVersion)
 		if err := autoupdate.CheckExecutable(context.TODO(), executableLocation(versionDir, binary), "--version"); err != nil {
-			ulm.removeUpdate(binary, versionsInLibrary[i].Original())
+			invalidVersions = append(invalidVersions, rawVersion)
 			continue
 		}
 
-		nonCurrentlyRunningVersionsKept += 1
+		versionsInLibrary = append(versionsInLibrary, v)
+	}
+
+	// Sort the versions (ascending order)
+	sort.Sort(semver.Collection(versionsInLibrary))
+
+	// Transform versions back into strings now that we've finished sorting them
+	versionsInLibraryStr := make([]string, len(versionsInLibrary))
+	for i, v := range versionsInLibrary {
+		versionsInLibraryStr[i] = v.Original()
 	}
+
+	return versionsInLibraryStr, invalidVersions, nil
+}
+
+// IsInstallVersion checks whether the version in the given target is the same as the one
+// contained in the installation.
+func (ulm *updateLibraryManager) IsInstallVersion(binary autoupdatableBinary, targetFilename string) bool {
```

Comment:
Hrm. So how will this get used?

1. installed as 1.0.8
2. upgraded to 1.0.9
3. downgraded to 1.0.8

So what is executed?

My gut sense is that the complexity isn't worth it, and just downloading 1.0.8 is fine

Reply:
Pulled it out, more trouble than it was worth!

Reply:
Having slept on it... I think we need _something_ here. 

There's a case that's like:
1. installed as 1.0.8
2. stable is 1.0.8
3. download 1.0.8

And that feels kinda -_-

Bt my inclination would be to push the initial versioning to `main` or potentially even the packaging layer.

Reply:
Main might be clean enough. I think the hardest part here was the find-binary code. And main / extension setup should know that

Reply:
I feel like I'm talking myself in circles here. I'm sorry the requirements aren't more clear.

Reply:
I've just wrapped up this commit https://github.com/kolide/launcher/pull/1178/commits/72db9509dd977bf2ed2e6c68b250805e9afa07d1, which I think handles what we're talking about:

1. install version is 1.0.8
1. current running version is therefore 1.0.8
1. there is no update downloaded for 1.0.8
1. we see that current running version is 1.0.8 so we don't perform the download